### ISSUE TITLE
NUnit2022 rule has false positive on hierarchies when using And

### ISSUE THREAD

**Issue by tupacko** [nunit/nunit.analyzers#306]
Hello,

it seems that the rule NUnit2022 has a false positive in case there are second level Prefixes (like another Property) applied. I say false positive, because with the analyzers not installed or disabled, the test runs as expected.

NUnit.Analyzers version: 0.5.0
NUnit: 3.12.0
Visual Studio 2019: 16.7.7

Given the following structure:

```
SomeClass
|- List<OtherClass>
   |- List<InnerListClass>
      |- LeafProperty
```

Given the assert for an instance of SomeClass:

```
Assert.That(
  someClass.OtherClasses,
  Has.Some.Property("InnerListClasses")
  .Some.TypeOf<SomeSpecificInnerListClass>()
  .And.Property("LeafProperty"));
```

*Eratta*: from before `TypeOf` a `Some` was missing in the original issue.

Expected: no error is raised by the analyzer, since `SomeClass` has a list of `OtherClass` instances, which in turn has a list of `InnerListClass` instances. The class `InnerListClass` has a property `LeafProperty`.
Actual: there is a compilation error raised because the analyzer says there is no `LeafProperty` on `IList<InnerListClass>`.

I also noted that in https://github.com/nunit/nunit.analyzers/blob/0.5.0/src/nunit.analyzers/MissingProperty/MissingPropertyAnalyzer.cs there is a comment on line 53. So the issue was considered when implementing, just not fully accommodated, I guess.

>  // Only first prefix supported (as preceding prefixes might change validated type)

Thank you!

**Comment by Dreamescaper** [comment]
I'll take this one

**Comment by Dreamescaper** [comment]
@tupacko 
Please provide example project.
I'm trying to reproduce your error, but the Assert itself fails for me.
```csharp
        [Test]
        public void ValidWhenPropertyChainIsUsed2()
        {
            var someClass = new SomeClass();
            Assert.That(someClass.OtherClasses,
              Has.Some.Property("InnerListClasses")
              .TypeOf<SomeSpecificInnerListClass>()
              .And.Property("LeafProperty"));
        }

        public class SomeClass
        {
            public List<OtherClass> OtherClasses { get; } = new List<OtherClass>
            {
                new OtherClass()
            };
        }

        public class OtherClass
        {
            public List<InnerListClass> InnerListClasses { get; } = new List<InnerListClass>
            {
                new SomeSpecificInnerListClass()
            };
        }

        public class InnerListClass
        {
            public string LeafProperty { get; set; }
        }

        public class SomeSpecificInnerListClass : InnerListClass
        {
        }
```

**Comment by Dreamescaper** [comment]
I've added a fix for issue with All/Some operators, but have no idea whether it's actually same issue.

**Comment by tupacko** [comment]
Hello,

I'll create an example project and post it here. Thank you for trying to fix the issue 👍 

Alpár

**Comment by tupacko** [comment]
Hello,

I edited the original assert, a `Some` was missing from before `TypeOf`. Also, I've created a test project with 4 tests exemplifying on the same data four different asserts. All are passing with NUnit 2022 disabled or without installing NUnit.Analyzers. Two of them are false positives in my understanding.

https://github.com/tupacko/NUnit2022HierarchyFalsePositive

Kind regards,
Alpár

**Comment by Dreamescaper** [comment]
So the fix that I've applied will help in those cases - simply because I'm disabling analyzer at all if Any/Some operator is present (we don't support it ATM).

There's another issue however. Problem is that 'LeafProperty' is not known at compile time.
So it would still trigger for something like this:
```csharp
InnerListClass innerListClass = new SomeSpecificInnerListClass();
Assert.That(innerListClass, Has.Property("LeafProperty").EqualTo("IDK"));
// ^ Diagnostic as InnerListClass has no LeafProperty prop.
```

@mikkelbu 
Do you think it makes sense to do anything about that? We could probably try finding possible inherited classes, and account for them as well.

**Comment by tupacko** [comment]
Maybe instead of error, the severity level could be warning if the class is not sealed. I mean, if `InnerListClass` would have been sealed, there is no chance of any derived class having (adding) that property.

I don't know what performance implication would it have to figure out whether the "runtime" instance will have/could have such a property. If at all possible at compile time.

**Comment by Dreamescaper** [comment]
It is possible to get all inherited types, and we already do that for some analyzers, but I'm not sure whether it makes any sense.
For example, one of the main use cases for this analyzer is to warn against `Has.Count` usage on `IEnumerable` actual argument. It won't work if we start looking at all inherited classes, as technically it might be `List` actual runtime type.

**Comment by mikkelbu** [comment]
> Do you think it makes sense to do anything about that? We could probably try finding possible inherited classes, and account for them as well.

I think for now that would be overkill and as you say this could lead to wrong results.
